boot/freeldr/freeldr/arch/uefi/uefildr.c

```diff
@@ -25,12 +28,50 @@ EfiEntry(
     GlobalImageHandle = ImageHandle;
     GlobalSystemTable = SystemTable;
 
-    BootMain(NULL);
+    /* Debugger pre-initialization */
+    DebugInit(0);
+
+    MachInit("");
+
+    /* UI pre-initialization */
+    if (!UiInitialize(FALSE))
+    {
+        UiMessageBoxCritical("Unable to initialize UI.");
+        //goto Quit;
+    }
+
+    /* Initialize memory manager */
+    if (!MmInitializeMemoryManager())
+    {
+        UiMessageBoxCritical("Unable to initialize memory manager.");
+        //goto Quit;
+    }
+
+    /* Initialize I/O subsystem */
+    FsInit();
+
+    /* 0x32000 is what UEFI defines, but we can go smaller if we want */
+    BasicStack = (PVOID)((ULONG_PTR)0x32000 + (ULONG_PTR)MmAllocateMemoryWithType(0x32000, LoaderLoadedProgram));
+    _changestack();
```

Comment:
We NEED to use the UEFI stack for every firmware call back it was way less janky to do this way.

Reply:
Here lemme be more detailed:
We call into a callback that needs the UEFI stack 5+ times before we call RunLoader();
afterword we need to call ExitBootServices

The idea here is that instead of calling that function directly we restore the UEFI stack and return to how we were in one assembly call.

This removes a lot of the instability, and ensures that we can modify uefimem.c later down the line.

From what i understand we can still use uefidisk.c on our new stack; As this is seemingly what GRUB and Limine do.